File: ios/GuideStreamTV/Services/SportsService.swift
//
//  SportsService.swift
//  GuideStreamTV
//
//  Fetches live + upcoming games from ESPN's public scoreboard endpoints.
//

import Foundation
import SwiftUI
import Supabase

// MARK: - Public Models

struct SportsGame: Identifiable, Hashable {
    let id: String
    let sport: String              // "NBA" / "NFL" / "Soccer" / "MLB" / "UFC"
    let leagueShort: String        // "NBA", "Premier League", etc
    let state: GameState
    let statusDetail: String       // "3rd Qtr · 8:42" or "8:30 PM ET" or "Final"
    let startDate: Date
    let home: GameTeam
    let away: GameTeam
    let broadcasts: [String]       // ["ESPN+", "TNT"]
}

enum GameState: String {
    case pre, live, post

    var isLive: Bool { self == .live }
}

struct GameTeam: Hashable {
    let id: String?
    let uid: String?
    let abbreviation: String
    let displayName: String
    let shortName: String
    let score: String
    let primaryHex: String?
    let isWinner: Bool
    var logoURL: String? = nil
}

// MARK: - Streaming Simulcast Companions

/// Deterministic linear-network → streaming-companion lookup for the sports
/// Where to Watch chips. ESPN's public scoreboard only reports the linear
/// carrier (e.g. "NBC") and never the streaming simulcast (e.g. Peacock), so
/// the chip row would otherwise never tell a cord-cutter where the game
/// actually streams. Applied ONLY at the point where the chips are built —
/// `SportsGame.broadcasts` itself is never mutated, so every other surface
/// (sports cards, detail screen, hero carousel, tvOS tiles) keeps showing
/// the linear network alone.
nonisolated enum SportsSimulcast {

    /// Normalized names that are already streaming destinations — these never
    /// get a companion appended.
    private static let streamingDestinations: Set<String> = [
        "peacock", "peacockpremium", "paramount", "paramountplus", "primevideo",
        "amazonprime", "netflix", "appletv", "appletvplus", "hbomax", "max",
        "hulu", "fubo", "fubotv", "slingtv", "youtube", "youtubetv", "tubi",
        "disneyplus", "espn", "espnplus", "nflplus", "nbaleaguepass", "mlbtv",
        "foxone"
    ]

    /// Exact normalized key → streaming companion. Matched exactly, never by
    /// substring, so unexpected network names (MSG, FDSSO, …) yield nothing.
    private static let companions: [String: String] = [
        "nbc": "Peacock", "nbcsn": "Peacock", "cnbc": "Peacock",
        "usa": "Peacock", "usanetwork": "Peacock", "usanet": "Peacock",
        "telemundo": "Peacock",
        "universo": "Peacock", "golf": "Peacock", "golfchannel": "Peacock",
        "cbs": "Paramount+", "cbssn": "Paramount+", "cbssportsnetwork": "Paramount+",
        "abc": "ESPN", "espn2": "ESPN", "espnu": "ESPN", "espnews": "ESPN",
        "espndeportes": "ESPN", "secn": "ESPN", "secnetwork": "ESPN",
        "secnplus": "ESPN",
        "accn": "ESPN", "accnetwork": "ESPN",
        "fox": "Fox One", "foxsports": "Fox One", "fs1": "Fox One",
        "fs2": "Fox One", "btn": "Fox One", "bigtennetwork": "Fox One",
        "foxdeportes": "Fox One",
        "tnt": "HBO Max", "tbs": "HBO Max", "trutv": "HBO Max",
        "nfln": "NFL+", "nflnetwork": "NFL+",
        "nbatv": "NBA League Pass",
        "mlbn": "MLB.TV", "mlbnetwork": "MLB.TV"
    ]

    /// Lowercased with every non-alphanumeric character removed.
    private static func normalize(_ name: String) -> String {
        name.lowercased().filter { $0.isLetter || $0.isNumber }
    }

    /// Returns the broadcasts with each linear network's streaming companion
    /// inserted immediately after it, deduplicated case-insensitively. Index
    /// zero is always the original first broadcast. These mappings encode US
    /// streaming rights only, so non-US regions get the input back unchanged
    /// rather than being pointed at a service that doesn't carry the game.
    static func enrich(_ broadcasts: [String]) -> [String] {
        guard DeviceLocale.current().region == "US" else { return broadcasts }
        var result: [String] = []
        var seen = Set<String>()
        for name in broadcasts {
            let key = normalize(name)
            if seen.insert(key).inserted {
                result.append(name)
            }
            guard !streamingDestinations.contains(key),
                  let companion = companions[key],
                  seen.insert(normalize(companion)).inserted else { continue }
            result.append(companion)
        }
        return result
    }
}

// MARK: - Service

@MainActor
final class SportsService {
    static let shared = SportsService()
    private init() {}

    private struct Endpoint {
        let sport: String
        let path: String
    }

    private let endpoints: [Endpoint] = [
        Endpoint(sport: "NBA",    path: "basketball/nba/scoreboard"),
        Endpoint(sport: "NBA Summer", path: "basketball/nba-summer-las-vegas/scoreboard"),
        Endpoint(sport: "NBA Summer", path: "basketball/nba-summer-utah/scoreboard"),
        Endpoint(sport: "NBA Summer", path: "basketball/nba-summer-sacramento/scoreboard"),
        Endpoint(sport: "NFL",    path: "football/nfl/scoreboard"),
        Endpoint(sport: "Soccer", path: "soccer/eng.1/scoreboard"),
        Endpoint(sport: "Soccer", path: "soccer/fifa.world/scoreboard"),
        Endpoint(sport: "MLB",    path: "baseball/mlb/scoreboard"),
        Endpoint(sport: "UFC",    path: "mma/ufc/scoreboard"),
        Endpoint(sport: "WNBA",   path: "basketball/wnba/scoreboard"),
        // groups=80 is ESPN's FBS group. Without it college-football returns
        // every division down to D3 (760 teams, 158 games on a Saturday), most
        // of which carry no national broadcast to deep-link into.
        Endpoint(sport: "CFB",    path: "football/college-football/scoreboard?groups=80"),
        Endpoint(sport: "NCAA Men",  path: "basketball/mens-college-basketball/scoreboard"),
        Endpoint(sport: "NCAA Women",  path: "basketball/womens-college-basketball/scoreboard")
    ]

    func fetchAll() async -> [SportsGame] {
        await withTaskGroup(of: [SportsGame].self) { group in
            for ep in endpoints {
                group.addTask { [weak self] in
                    guard let self else { return [] }
                    return await self.fetch(endpoint: ep)
                }
            }
            var all: [SportsGame] = []
            for await games in group { all.append(contentsOf: games) }
            return all.sorted { a, b in
                if a.state.isLive != b.state.isLive { return a.state.isLive }
                return a.startDate < b.startDate
            }
        }
    }

    /// Single-game refresh used by the game detail screen. Re-fetches only the
    /// endpoints whose sport matches `game.sport` (1 call for MLB/NBA/NFL/
    /// Soccer/UFC, 3 for "NBA Summer") and returns the game with the same id,
    /// or nil when it is no longer on the scoreboard / every call failed.
    /// Reuses `fetch(endpoint:)` and its mapping — no second ESPN parser.
    func refresh(game: SportsGame) async -> SportsGame? {
        let matching = endpoints.filter { $0.sport == game.sport }
        guard !matching.isEmpty else { return nil }
        return await withTaskGroup(of: [SportsGame].self) { group -> SportsGame? in
            for ep in matching {
                group.addTask { [weak self] in
                    guard let self else { return [] }
                    return await self.fetch(endpoint: ep)
                }
            }
            var match: SportsGame?
            for await games in group where match == nil {
                match = games.first { $0.id == game.id }
            }
            return match
        }
    }

    /// Resolves a single game by its ESPN id from Supabase's `sports_games`
    /// table, which `sports_poll_and_notify` keeps current.
    ///
    /// `fetchAll()` reads ESPN's live scoreboards directly, so it only ever
    /// contains games on today's slate — a "Final" push tapped later, or any
    /// tap made while ESPN is refusing the app's requests, finds nothing there
    /// and the sports push silently opens no detail. `sports_games` always
    /// holds the row the push was generated from, so this is the authoritative
    /// lookup for a notification tap. Returns nil when the id is unknown.
    func fetchGame(id gameId: String) async -> SportsGame? {
        let trimmed = gameId.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return nil }
        do {
            let rows: [SportsGameRow] = try await SupabaseManager.shared.client
                .from("sports_games")
                .select()
                .eq("game_id", value: trimmed)
                .limit(1)
                .execute()
                .value
            return rows.first.map(Self.mapRow)
        } catch {
            print("[SportsService] fetchGame(\(trimmed)) failed: \(error.localizedDescription)")
            return nil
        }
    }

    /// Builds a `SportsGame` from a `sports_games` row. Scores are stored as
    /// integers there and as display strings on the model, so nil becomes "".
    private static func mapRow(_ r: SportsGameRow) -> SportsGame {
        let state: GameState = {
            switch r.state {
            case "live": return .live
            case "final": return .post
            default: return .pre
            }
        }()
        let homeScore = r.home_score.map(String.init) ?? ""
        let awayScore = r.away_score.map(String.init) ?? ""
        let homeWins = (r.home_score ?? 0) > (r.away_score ?? 0)
        return SportsGame(
            id: r.game_id,
            sport: r.sport ?? "",
            leagueShort: r.sport ?? "",
            state: state,
            statusDetail: r.status_detail ?? "",
            startDate: r.start_at.flatMap(parseDate) ?? Date(),
            home: GameTeam(
                id: r.home_id,
                uid: r.home_uid,
                abbreviation: r.home_abbr ?? "",
                displayName: r.home_name ?? "",
                shortName: r.home_name ?? "",
                score: homeScore,
                primaryHex: nil,
                isWinner: state == .post && homeWins
            ),
            away: GameTeam(
                id: r.away_id,
                uid: r.away_uid,
                abbreviation: r.away_abbr ?? "",
                displayName: r.away_name ?? "",
                shortName: r.away_name ?? "",
                score: awayScore,
                primaryHex: nil,
                isWinner: state == .post && !homeWins
            ),
            broadcasts: [r.broadcast].compactMap { $0 }.filter { !$0.isEmpty }
        )
    }

    private func fetch(endpoint ep: Endpoint) async -> [SportsGame] {
        let urlString = "https://site.api.espn.com/apis/site/v2/sports/\(ep.path)"
        guard let url = URL(string: urlString) else { return [] }
        do {
            let (data, _) = try await URLSession.shared.data(from: url)
            let decoded = try JSONDecoder().decode(ESPNScoreboard.self, from: data)
            return decoded.events?.compactMap { ev in
                Self.mapEvent(ev, sport: ep.sport)
            } ?? []
        } catch {
            print("[SportsService] \(ep.sport) fetch failed: \(error)")
            return []
        }
    }

    private static func mapEvent(_ ev: ESPNEvent, sport: String) -> SportsGame? {
        guard let comp = ev.competitions?.first,
              let competitors = comp.competitors, competitors.count >= 2 else { return nil }

        let homeRaw = competitors.first(where: { $0.homeAway == "home" }) ?? competitors[0]
        let awayRaw = competitors.first(where: { $0.homeAway == "away" }) ?? competitors[1]

        let state: GameState = {
            switch ev.status?.type?.state {
            case "in": return .live
            case "post": return .post
            default: return .pre
            }
        }()

        let detail: String = {
            if state == .live, let s = ev.status?.type?.shortDetail { return s }
            if state == .post { return ev.status?.type?.shortDetail ?? "Final" }
            // pre — format like "8:30 PM ET"
            if let date = parseDate(ev.date) {
                let f = DateFormatter()
                f.dateFormat = "h:mm a"
                f.timeZone = TimeZone(identifier: "America/New_York")
                return "\(f.string(from: date)) ET"
            }
            return ev.status?.type?.shortDetail ?? ""
        }()

        let broadcasts: [String] = {
            let names = comp.broadcasts?.flatMap { $0.names ?? [] } ?? []
            return Array(Set(names)).sorted()
        }()

        let leagueShort = ev.season?.slug ?? sport

        return SportsGame(
            id: ev.id ?? UUID().uuidString,
            sport: sport,
            leagueShort: leagueShort,
            state: state,
            statusDetail: detail,
            startDate: parseDate(ev.date) ?? .distantPast,
            home: makeTeam(from: homeRaw),
            away: makeTeam(from: awayRaw),
            broadcasts: broadcasts
        )
    }

    private static func makeTeam(from c: ESPNCompetitor) -> GameTeam {
        let fallbackAbbrev: String = {
            if let s = c.team?.shortDisplayName { return String(s.prefix(3)).uppercased() }
            return "—"
        }()
        return GameTeam(
            id: c.team?.id,
            uid: c.team?.uid,
            abbreviation: c.team?.abbreviation ?? fallbackAbbrev,
            displayName: c.team?.displayName ?? "—",
            shortName: c.team?.shortDisplayName ?? c.team?.name ?? "—",
            score: c.score ?? "0",
            primaryHex: c.team?.color,
            isWinner: c.winner ?? false,
            logoURL: c.team?.logo
        )
    }

    private static func parseDate(_ s: String?) -> Date? {
        guard let s else { return nil }

        // ESPN scoreboard dates come in several variants:
        //   2026-06-27T00:00:00.000Z  (fractional seconds)
        //   2026-06-27T00:00:00Z     (with seconds)
        //   2026-06-27T00:00Z        (no seconds — most common)
        // Try strict ISO8601 first, then fall back to custom formats.

        let isoFractional = ISO8601DateFormatter()
        isoFractional.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let d = isoFractional.date(from: s) { return d }

        let iso = ISO8601DateFormatter()
        iso.formatOptions = [.withInternetDateTime]
        if let d = iso.date(from: s) { return d }

        let withSeconds: DateFormatter = {
            let f = DateFormatter()
            f.locale = Locale(identifier: "en_US_POSIX")
            f.dateFormat = "yyyy-MM-dd'T'HH:mm:ssXXXXX"
            return f
        }()
        if let d = withSeconds.date(from: s) { return d }

        let withoutSeconds: DateFormatter = {
            let f = DateFormatter()
            f.locale = Locale(identifier: "en_US_POSIX")
            f.dateFormat = "yyyy-MM-dd'T'HH:mmXXXXX"
            return f
        }()
        if let d = withoutSeconds.date(from: s) { return d }

        return nil
    }
}

// MARK: - ESPN response models (nonisolated, decoded on background)

nonisolated struct ESPNScoreboard: Decodable {
    let events: [ESPNEvent]?
}

nonisolated struct ESPNEvent: Decodable {
    let id: String?
    let date: String?
    let status: ESPNStatus?
    let competitions: [ESPNCompetition]?
    let season: ESPNSeason?
}

nonisolated struct ESPNSeason: Decodable {
    let slug: String?
}

nonisolated struct ESPNStatus: Decodable {
    let type: ESPNStatusType?
}

nonisolated struct ESPNStatusType: Decodable {
    let state: String?
    let shortDetail: String?
    let completed: Bool?
}

nonisolated struct ESPNCompetition: Decodable {
    let competitors: [ESPNCompetitor]?
    let broadcasts: [ESPNBroadcast]?
}

nonisolated struct ESPNBroadcast: Decodable {
    let names: [String]?
}

nonisolated struct ESPNCompetitor: Decodable {
    let homeAway: String?
    let score: String?
    let winner: Bool?
    let team: ESPNTeam?
}

nonisolated struct ESPNTeam: Decodable {
    let id: String?
    let uid: String?
    let abbreviation: String?
    let displayName: String?
    let shortDisplayName: String?
    let name: String?
    let color: String?
    let logo: String?
}


/// Row decoder for `public.sports_games`, written by the
/// `sports_poll_and_notify` edge function. Every column but `game_id` is
/// optional so a partially-populated row still decodes.
nonisolated struct SportsGameRow: Decodable, Sendable {
    let game_id: String
    let league: String?
    let sport: String?
    let home_uid: String?
    let home_id: String?
    let home_abbr: String?
    let home_name: String?
    let away_uid: String?
    let away_id: String?
    let away_abbr: String?
    let away_name: String?
    let state: String?
    let status_detail: String?
    let home_score: Int?
    let away_score: Int?
    let start_at: String?
    let broadcast: String?
}

File: android/app/src/main/java/com/rork/guidestreamtvandroid/data/remote/SportsService.kt
package com.rork.guidestreamtvandroid.data.remote

import android.util.Log
import com.rork.guidestreamtvandroid.data.models.SportsGame
import io.github.jan.supabase.postgrest.postgrest
import io.ktor.client.HttpClient
import io.ktor.client.call.body
import io.ktor.client.plugins.DefaultRequest
import io.ktor.client.plugins.contentnegotiation.ContentNegotiation
import io.ktor.client.request.get
import io.ktor.client.request.header
import io.ktor.http.HttpHeaders
import io.ktor.serialization.kotlinx.json.json
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.withContext
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json

/**
 * Sports service — mirrors iOS SportsService.swift.
 * Fetches live + upcoming games from ESPN's public scoreboard endpoints.
 */
class SportsService {

    private val client = HttpClient {
        // ESPN's CDN (Akamai) 403s requests whose User-Agent doesn't match the
        // client's fingerprint. Ktor's Android engine sends no UA by default,
        // so every scoreboard request returned 403 with an HTML error page
        // (which then broke JSON parsing downstream). The engine is
        // HttpURLConnection, which is backed by OkHttp internally, so an
        // okhttp UA is fingerprint-coherent and verified accepted by ESPN;
        // desktop browser agent strings get blocked on fingerprint mismatch.
        install(DefaultRequest) {
            header(HttpHeaders.UserAgent, "okhttp/4.12.0")
            header(HttpHeaders.Accept, "application/json")
        }
        install(ContentNegotiation) {
            json(Json { ignoreUnknownKeys = true; coerceInputValues = true; isLenient = true })
        }
    }

    private data class Endpoint(val sport: String, val path: String)

    private val endpoints = listOf(
        Endpoint("NBA", "basketball/nba/scoreboard"),
        Endpoint("NBA Summer", "basketball/nba-summer-las-vegas/scoreboard"),
        Endpoint("NBA Summer", "basketball/nba-summer-utah/scoreboard"),
        Endpoint("NBA Summer", "basketball/nba-summer-sacramento/scoreboard"),
        Endpoint("NFL", "football/nfl/scoreboard"),
        Endpoint("Soccer", "soccer/eng.1/scoreboard"),
        Endpoint("Soccer", "soccer/fifa.world/scoreboard"),
        Endpoint("MLB", "baseball/mlb/scoreboard"),
        Endpoint("UFC", "mma/ufc/scoreboard"),
    )

    @Serializable
    private data class ESPNResponse(
        val events: List<ESPNEvent> = emptyList(),
    )

    @Serializable
    private data class ESPNEvent(
        val id: String = "",
        val date: String = "",
        val status: ESPNStatus? = null,
        val competitions: List<ESPNCompetition> = emptyList(),
        val season: ESPNSeason? = null,
    )

    @Serializable
    private data class ESPNStatus(
        val type: ESPNStatusType? = null,
    )

    @Serializable
    private data class ESPNStatusType(
        val state: String = "pre",
        val shortDetail: String = "",
    )

    @Serializable
    private data class ESPNCompetition(
        val competitors: List<ESPNCompetitor> = emptyList(),
        val broadcasts: List<ESPNBroadcast> = emptyList(),
    )

    @Serializable
    private data class ESPNCompetitor(
        val id: String? = null,
        val uid: String? = null,
        val score: String = "",
        @SerialName("homeAway") val homeAway: String = "home",
        val team: ESPNTeam? = null,
        val winner: Boolean? = null,
    )

    @Serializable
    private data class ESPNTeam(
        val id: String? = null,
        val uid: String? = null,
        val abbreviation: String = "",
        @SerialName("displayName") val displayName: String = "",
        @SerialName("shortDisplayName") val shortDisplayName: String = "",
        val name: String = "",
        val color: String? = null,
        val logo: String? = null,
    )

    @Serializable
    private data class ESPNBroadcast(
        val names: List<String> = emptyList(),
    )

    @Serializable
    private data class ESPNSeason(
        val slug: String? = null,
    )

    /**
     * Row decoder for `public.sports_games`, written by the
     * `sports_poll_and_notify` edge function. Every column but `game_id` is
     * optional so a partially-populated row still decodes.
     */
    @Serializable
    private data class SportsGameRow(
        @SerialName("game_id") val gameId: String,
        val league: String? = null,
        val sport: String? = null,
        @SerialName("home_uid") val homeUid: String? = null,
        @SerialName("home_id") val homeId: String? = null,
        @SerialName("home_abbr") val homeAbbr: String? = null,
        @SerialName("home_name") val homeName: String? = null,
        @SerialName("away_uid") val awayUid: String? = null,
        @SerialName("away_id") val awayId: String? = null,
        @SerialName("away_abbr") val awayAbbr: String? = null,
        @SerialName("away_name") val awayName: String? = null,
        val state: String? = null,
        @SerialName("status_detail") val statusDetail: String? = null,
        @SerialName("home_score") val homeScore: Int? = null,
        @SerialName("away_score") val awayScore: Int? = null,
        @SerialName("start_at") val startAt: String? = null,
        val broadcast: String? = null,
    )

    /** Fetch all games across all sports, sorted live-first then by start time. */
    suspend fun fetchAll(): List<SportsGame> = withContext(Dispatchers.IO) {
        coroutineScope {
            val results = endpoints.map { ep ->
                async { fetch(ep) }
            }.awaitAll()
            val all = results.flatten()
            all.sortedWith(compareByDescending<SportsGame> { it.state == "live" }.thenBy { it.startDate ?: Long.MAX_VALUE })
        }
    }

    /**
     * Resolves a single game by its ESPN id from Supabase's `sports_games`
     * table, which the `sports_poll_and_notify` edge function keeps current.
     *
     * [fetchAll] reads ESPN's live scoreboards directly, so it only ever holds
     * today's slate — a "Final" push tapped later, or any tap made while ESPN
     * is refusing the request, finds nothing there and the notification opens
     * no detail sheet (GUI-46). `sports_games` always holds the row the push
     * was generated from, so it is the authoritative lookup for a tap. Returns
     * null when the id is genuinely unknown.
     */
    suspend fun fetchGame(gameId: String): SportsGame? = withContext(Dispatchers.IO) {
        val trimmed = gameId.trim()
        if (trimmed.isEmpty()) return@withContext null
        try {
            SupabaseManager.client.postgrest
                .from("sports_games")
                .select { filter { eq("game_id", trimmed) }; limit(1) }
                .decodeList<SportsGameRow>()
                .firstOrNull()
                ?.let(::mapRow)
        } catch (e: Exception) {
            Log.w("SportsService", "fetchGame($trimmed) failed", e)
            null
        }
    }

    /**
     * Builds a [SportsGame] from a `sports_games` row. The table stores state
     * as "pre" | "live" | "final" while the model uses ESPN's "pre" | "live" |
     * "post", so "final" is normalized here.
     */
    private fun mapRow(r: SportsGameRow): SportsGame {
        val state = when (r.state) {
            "live" -> "live"
            "final", "post" -> "post"
            else -> "pre"
        }
        val start = parseDate(r.startAt)
        val statusDetail = r.statusDetail?.takeIf { it.isNotBlank() }
            ?: when (state) {
                "post" -> "Final"
                else -> start?.let { formatGameTime(it) } ?: ""
            }
        val homeWins = (r.homeScore ?: 0) > (r.awayScore ?: 0)
        val sport = r.sport.orEmpty()
        return SportsGame(
            id = r.gameId,
            sport = sport,
            leagueShort = sport,
            state = state,
            statusDetail = statusDetail,
            home = SportsGame.TeamSummary(
                name = r.homeName ?: r.homeAbbr.orEmpty(),
                abbreviation = r.homeAbbr.orEmpty(),
                logoUrl = null,
                record = null,
                uid = r.homeUid ?: r.homeId,
                displayName = r.homeName ?: r.homeAbbr.orEmpty(),
                shortName = r.homeName ?: r.homeAbbr.orEmpty(),
                score = r.homeScore?.toString() ?: "",
                primaryHex = null,
                isWinner = state == "post" && homeWins,
            ),
            away = SportsGame.TeamSummary(
                name = r.awayName ?: r.awayAbbr.orEmpty(),
                abbreviation = r.awayAbbr.orEmpty(),
                logoUrl = null,
                record = null,
                uid = r.awayUid ?: r.awayId,
                displayName = r.awayName ?: r.awayAbbr.orEmpty(),
                shortName = r.awayName ?: r.awayAbbr.orEmpty(),
                score = r.awayScore?.toString() ?: "",
                primaryHex = null,
                isWinner = state == "post" && !homeWins,
            ),
            startTime = r.startAt,
            startDate = start?.toEpochMilli(),
            broadcasts = listOfNotNull(r.broadcast?.takeIf { it.isNotBlank() }),
            homeScore = r.homeScore,
            awayScore = r.awayScore,
        )
    }

    private suspend fun fetch(endpoint: Endpoint): List<SportsGame> {
        return try {
            val url = "https://site.api.espn.com/apis/site/v2/sports/${endpoint.path}"
            val response: ESPNResponse = client.get(url).body()
            response.events.mapNotNull { ev ->
                val comp = ev.competitions.firstOrNull() ?: return@mapNotNull null
                val home = comp.competitors.find { it.homeAway == "home" } ?: return@mapNotNull null
                val away = comp.competitors.find { it.homeAway == "away" } ?: return@mapNotNull null
                // ESPN reports state as "pre" | "in" | "post"; normalize "in" to "live".
                val rawState = ev.status?.type?.state ?: "pre"
                val state = if (rawState == "in") "live" else rawState
                val detail = ev.status?.type?.shortDetail ?: ""
                val startDate = parseDate(ev.date)
                val statusDetail = when (state) {
                    "live" -> detail
                    "post" -> detail.ifEmpty { "Final" }
                    else -> startDate?.let { formatGameTime(it) } ?: detail
                }
                val broadcasts = comp.broadcasts.flatMap { it.names }
                SportsGame(
                    id = ev.id,
                    sport = endpoint.sport,
                    leagueShort = ev.season?.slug ?: endpoint.sport,
                    state = state,
                    statusDetail = statusDetail,
                    home = makeTeam(home),
                    away = makeTeam(away),
                    startTime = ev.date,
                    startDate = startDate?.toEpochMilli(),
                    broadcasts = broadcasts,
                    homeScore = home.score.toIntOrNull(),
                    awayScore = away.score.toIntOrNull(),
                )
            }
        } catch (e: Exception) {
            Log.w("SportsService", "Fetch failed for ${endpoint.path}", e)
            emptyList()
        }
    }

    private fun makeTeam(c: ESPNCompetitor): SportsGame.TeamSummary {
        val team = c.team
        val fallbackAbbrev = team?.shortDisplayName?.takeIf { it.isNotBlank() }?.take(3)?.uppercase() ?: "—"
        val abbreviation = team?.abbreviation?.takeIf { it.isNotBlank() } ?: fallbackAbbrev
        val displayName = team?.displayName?.takeIf { it.isNotBlank() }
            ?: team?.name?.takeIf { it.isNotBlank() }
            ?: abbreviation
        val shortName = team?.shortDisplayName?.takeIf { it.isNotBlank() }
            ?: team?.name?.takeIf { it.isNotBlank() }
            ?: abbreviation
        return SportsGame.TeamSummary(
            name = displayName,
            abbreviation = abbreviation,
            logoUrl = team?.logo,
            record = null,
            uid = team?.uid ?: c.uid,
            displayName = displayName,
            shortName = shortName,
            score = c.score,
            primaryHex = team?.color,
            isWinner = c.winner ?: false,
        )
    }

    private fun parseDate(s: String?): java.time.Instant? {
        if (s.isNullOrBlank()) return null
        return try {
            java.time.Instant.parse(s)
        } catch (_: Exception) {
            try {
                java.time.OffsetDateTime.parse(s).toInstant()
            } catch (_: Exception) {
                null
            }
        }
    }

    private fun formatGameTime(instant: java.time.Instant): String {
        val formatter = java.time.format.DateTimeFormatter.ofPattern("h:mm a")
            .withZone(java.time.ZoneId.of("America/New_York"))
        return "${formatter.format(instant)} ET"
    }

    companion object {
        @Volatile private var instance: SportsService? = null
        fun get(): SportsService = instance ?: synchronized(this) {
            instance ?: SportsService().also { instance = it }
        }
    }
}
